keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,175 @@
+    (addition/delete is not recursive).
+  - For associative lists, keys are already "immutable" in a way (similar to how
+    names are "immutables" for objects). Changing the keys would mean deleting
+    and creating a new item in the list, which would be permitted by these
+    rules.
+- Mutable: A field set as mutable could cancel the recursive immutability.
+
+#### On mutation of selected fields
+
+- Error: mutation of an immutable field returns an error and the request is
+  rejected. No mutation is performed at all.
+- Ignored: mutation of the field (and/or sub-fields) are completely ignored, and
+  not persisted. They do not trigger an error.
+
+#### Proposal
+
+The behavior on mutation of immutable field is orthogonal to selection of
+immutable fields. If we accepted these two semantics at the same time, we would
+have to either have two different markers (one for selection and one for
+behavior) or have a single field with multiple options.
+
+For the moment, we understand that "Recursive" is the most important and simple
+selection mechanism, and "error" is the most natural way to handle mutations.
+
+### Go IDL
+
+The proposal is to add a new marker with the following (only) format:
+
+```
+// The name can not be changed after creation.
+// +immutable
+Name string
+
+// The list of containers can not change AT ALL after creation.
+// No single field in existing containers can be changed, added or deleted,
+// no new containers can be added, no existing container can be removed.
+// +immutable
+Containers []Containers
+```
+
+### OpenAPI Extension
+
+Since the semantics of the "readOnly" tag in OpenAPI is not the one we're trying
+to have here.  We propose a new Kubernetes specific extensions, which has to
+allow for further changes in the semantics of our immutable marker:
+```
+X-Kubernetes-immutable: recursive,error
+```
+
+### Mutating admission chain
+Mutating admission chain would have the exact same effects as user changes,
+meaning that they wouldn't be able to change an object after creation. This is
+very similar to what is done today since validation for update is run AFTER all
+mutations. On creation, mutation are permitted.
+
+### Risks and Mitigations
+
+The risks are considered low. For CRDs, it would be possible for someone to set
+immutable fields on metadata, which could result in objects that can't be
+deleted or updated ever. For example, having finalizers, deletionTimestamp or
+managedFields set to immutable would make the object unusable.
+
+We can mitigate that by forbidding any of these annotations on metadata fields.
+
+## Design Details
+
+### Test Plan
+
+Not ready yet
+
+### Graduation Criteria
+
+This feature is mostly considered as a sub-feature of [Server-side
+apply](https://github.com/kubernetes/enhancement/pull/555) and will probably
+follow a very similar graduation line. It is expected to start directly as a
+beta feature and will graduate to GA along with Server-side apply.
```

Comment:
why is it a subfeature? Other than shared code, I don't see a strong connection to server side apply. I wouldn't connect them.

Reply:
Well, we need the OpenAPI schema for both CRDs and native type. We already have the logic in place in SSA to handle both.

Reply:
Am not speaking against using the SSA code, just that this is a different feature.

Reply:
I'll rephrase then